streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala

```diff
@@ -17,29 +17,791 @@ class SinkSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def orElseExtractErrorBoth = {
+    val sink = extractErrorSink orElse extractErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def raceBothLeft = {
+    val sink = ZSink.identity[Int] raceBoth ZSink.succeedLazy("Hello")
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def raceBothInitErrorLeft = {
+    val sink = initErrorSink raceBoth ZSink.identity[Int]
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Right(1)))
+  }
+
+  private def raceBothInitErrorRight = {
+    val sink = ZSink.identity[Int] raceBoth initErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def raceBothInitErrorBoth = {
+    val sink = initErrorSink race initErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def raceBothStepErrorLeft = {
+    val sink = stepErrorSink raceBoth ZSink.identity[Int]
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def raceBothStepErrorRight = {
+    val sink = ZSink.identity[Int] raceBoth stepErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def raceBothStepErrorBoth = {
+    val sink = stepErrorSink race stepErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def raceBothExtractErrorLeft = {
+    val sink = extractErrorSink raceBoth ZSink.identity[Int]
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def raceBothExtractErrorRight = {
+    val sink = ZSink.identity[Int] raceBoth extractErrorSink
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Left(1)))
+  }
+
+  private def raceBothExtractErrorBoth = {
+    val sink = extractErrorSink race extractErrorSink
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def takeWhileHappyPath = {
+    val sink = ZSink.identity[Int].takeWhile[Int](_ < 5)
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== 1))
+  }
+
+  private def takeWhileFalsePredicate = {
+    val sink = ZSink.identity[Int].takeWhile[Int](_ > 5)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left(())))
+  }
+
+  private def takeWhileInitError = {
+    val sink = initErrorSink.takeWhile[Int](_ < 5)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def takeWhileStepError = {
+    val sink = stepErrorSink.takeWhile[Int](_ < 5)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def takeWhileExtractError = {
+    val sink = extractErrorSink.takeWhile[Int](_ < 5)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def untilOutputHappyPath = {
+    // This test needs to be verified for correctness.
```

Comment:
The logic in `untilOutput`'s step looks backwards. should be `if (!Step.done(s)) ...`